File: interview/exist.go
package interview

func exist(board [][]byte, word string) bool {
	if board == nil {
		return false
	}
	if word == "" {
		return true
	}
	memo := make([][]bool, len(board))
	for i := range memo {
		memo[i] = make([]bool, len(board[0]))
	}
	var dfs func(r, c, i int) bool
	dfs = func(r, c, i int) bool {
		if i == len(word) {
			return true
		}
		flag := false
		memo[r][c] = true
		if r+1 < len(board) && !memo[r+1][c] && board[r+1][c] == word[i] {
			flag = dfs(r+1, c, i+1)
		}
		if r > 0 && !memo[r-1][c] && board[r-1][c] == word[i] && !flag {
			flag = dfs(r-1, c, i+1)
		}
		if c+1 < len(board[0]) && !memo[r][c+1] && board[r][c+1] == word[i] && !flag {
			flag = dfs(r, c+1, i+1)
		}
		if c > 0 && !memo[r][c-1] && board[r][c-1] == word[i] && !flag {
			flag = dfs(r, c-1, i+1)
		}
		memo[r][c] = false
		return flag
	}
	for row, ra := range board {
		for col, ca := range ra {
			if ca == word[0] {
				if dfs(row, col, 1) {
					return true
				}
			}
		}
	}
	return false
}

func T(s string, b byte) bool {
	if s[0] == b {
		return true
	}
	return false
}

File: host100/Exist.java
/**
 * @author : zexho
 * @created : 2022-05-29
**/
public class Exist{
    public boolean exist(char[][] board, String word) {
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                if(process(board,word,0,i,j)) return true;
            }
        }
        return false;
    }

    public boolean process(char[][] board,String word,int idx,int row,int col){
        if(idx == word.length()) return true;
        if(row < 0 || row == board.length) return false;
        if(col < 0 || col == board[0].length) return false;
        if(board[row][col] == '0') return false;
        if(board[row][col] != word.charAt(idx)) return false;

        char t = board[row][col];
        board[row][col] = '0';
        boolean res = process(board,word,idx+1,row-1,col) || process(board,word,idx+1,row,col-1)
                || process(board,word,idx+1,row+1,col) || process(board,word,idx+1,row,col+1);
        board[row][col] = t;
        return res;
    }
}
